river/anomaly/xstream.py

```diff
@@ -0,0 +1,220 @@
+    Args:
+        n_components (int): The number of components for streamhash projection (Default=100).
+        n_chains (int): The number of half-space chains (Default=100).
+        depth (int): The maximum depth for the chains (Default=25).
+        window_size (int): The size (and the sliding length) of the reference window (Default=25).
+    """
+
+    def __init__(
+            self,
+            num_components=100,
+            n_chains=100,
+            depth=25,
+            window_size=25):
+        self.streamhash = StreamhashProjector(num_components=num_components)
+        deltamax = np.ones(num_components) * 0.5
+        deltamax[np.abs(deltamax) <= 0.0001] = 1.0
+        self.window_size = window_size
+        self.hs_chains = _HSChains(
+            deltamax=deltamax,
+            n_chains=n_chains,
+            depth=depth)
+
+        self.step = 0
+        self.cur_window = []
+        self.ref_window = None
+
+    def learn_one(self, x, y=None):
+        """Fits the model to next instance.
+        Args:
+            X (np.float array of shape (num_features,)): The instance to fit.
+            y (int): Ignored since the model is unsupervised (Default=None).
+        Returns:
+            object: Returns the self.
+        """
+        self.step += 1
+
+        #X = self.streamhash.learn_transform_one(X)
+        x_array = dict2numpy(x)
+        x_array = self.streamhash.learn_transform_one(x_array)
+
+        #X = X.reshape(1, -1)
+        x_array = x_array.reshape(1,-1)
+        self.cur_window.append(x_array)
+
+        self.hs_chains.fit(x_array)
+
+        if self.step % self.window_size == 0:
+            self.ref_window = self.cur_window
+            self.cur_window = []
+            deltamax = self._compute_deltamax()
+            self.hs_chains.set_deltamax(deltamax)
+            self.hs_chains.next_window()
+
+        return self
+
+    def score_one(self, x):
+        """Scores the anomalousness of the next instance.
+        Args:
+            X (np.float array of shape (num_features,)): The instance to score. Higher scores represent more anomalous instances whereas lower scores correspond to more normal instances.
+        Returns:
+            score (float): The anomalousness score of the input instance.
+        """
+        x_array = dict2numpy(x)
+        x_array = self.streamhash.learn_transform_one(x_array)
+        #X = X.reshape(1, -1)
+        x_array = x_array.reshape(1,-1)
+        score = self.hs_chains.score(x_array).flatten()
+
+        return score
+
+    def _compute_deltamax(self):
+        # mx = np.max(np.concatenate(self.ref_window, axis=0), axis=0)
+        # mn = np.min(np.concatenate(self.ref_window, axis=0), axis=0)
+        mn, mx = get_minmax_array(np.concatenate(self.ref_window, axis=0))
+
+        deltamax = (mx - mn) / 2.0
+        deltamax[np.abs(deltamax) <= 0.0001] = 1.0
+
+        return deltamax
+
+
+class _Chain:
```

Comment:
Can you say what this class does? It's a bit difficult to understand from just looking at the code